src/BenchmarksApps/Kestrel/PlatformBenchmarks/BenchmarkApplication.HttpConnection.cs

```diff
@@ -41,41 +41,50 @@ private async Task ProcessRequestsAsync()
         {
             while (true)
             {
-                var task = Reader.ReadAsync();
-
-                if (!task.IsCompleted)
+                var readResult = await Reader.ReadAsync();
+                
+                if (!HandleRequest(readResult))
                 {
-                    // No more data in the input
-                    await OnReadCompletedAsync();
```

Comment:
Is there any concern that removing this logic this will hurt pipelined performance? Now there's at least once call to Writer.FlushAsync() per read.

Reply:
> Is there any concern that removing this logic this will hurt pipelined performance?

I've run both JSON and Plaintext benchmarks and it won't.

> Now there's at least one call to Writer.FlushAsync() per read.

So far there was also one but in a less obvious way: when there was no input data ready, we were flushing the output. 

With this change we always do it in explicit way: read request, handle it and write the response.

Reply:
> So far there was also one but in a less obvious way: when there was no input data ready, we were flushing the output.

That was to batch the 16 requests in the smallest number of TCP packets? Flushing on each request would fragment the response across many packets?

Reply:
Please excuse me for the delayed answer, I had to switch to Linux and attach a debugger to check the exact request sizes.

The pipelining is transparent to the web server.  

When it's disabled, we receive 49 bytes from `Reader.ReadAsync`.
When it's enabled, we receive 49 * 16 bytes from `Reader.ReadAsync`

The point is that so far, we were calling `GetWriter` and `GetSpan` 16 times for Plaintext in the body loop, now we do it just once, hence the 200k RPS gain.

When it comes to flushing, the behaviour is the same, but the code is different. So far we were flushing when there was no input data available, which was when server has sent a request, but did not receive the response yet. With my change we do it in more explicit way. As I said, I've run both Plaintext and Json platform benchmarks and they did not regress.

Reply:
Ah because `HandleRequest` processes multiple?

Reply:
yes, exactly